compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/Draggable.skiko.kt

```diff
@@ -0,0 +1,207 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.foundation
+
+import androidx.compose.foundation.gestures.awaitPointerSlopOrCancellation
+import androidx.compose.foundation.gestures.drag
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.input.pointer.AwaitPointerEventScope
+import androidx.compose.ui.input.pointer.PointerButton
+import androidx.compose.ui.input.pointer.PointerEvent
+import androidx.compose.ui.input.pointer.PointerEventType
+import androidx.compose.ui.input.pointer.PointerInputChange
+import androidx.compose.ui.input.pointer.PointerKeyboardModifiers
+import androidx.compose.ui.input.pointer.PointerType
+import androidx.compose.ui.input.pointer.changedToUpIgnoreConsumed
+import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.input.pointer.positionChange
+import androidx.compose.ui.platform.LocalKeyboardModifiers
+import androidx.compose.ui.util.fastAll
+import kotlin.jvm.JvmInline
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.isActive
+
+@ExperimentalFoundationApi
+data class DragChange(
+    val offset: Offset = Offset.Zero,
+    val previousKeyboardModifiers: PointerKeyboardModifiers,
+    val currentKeyboardModifiers: PointerKeyboardModifiers
+) {
+    fun keyboardModifierChanged(): Boolean =
+        previousKeyboardModifiers != currentKeyboardModifiers
+}
+
+
+@JvmInline
+value class DragStartResult(val dragAmount: Offset = Offset.Zero)
+
+@OptIn(ExperimentalFoundationApi::class)
+typealias PointerFilter = PointerFilterScope.() -> Boolean
+
+@Suppress("MemberVisibilityCanBePrivate")
+object AwaitDragStart {
+    @OptIn(ExperimentalFoundationApi::class)
+    val Default: suspend AwaitPointerEventScope.(PointerEvent, PointerFilter) -> DragStartResult? =
+        { initialDown, filter ->
+            val awaitVariant = when (initialDown.changes[0].type) {
+                PointerType.Mouse -> OnSlop
+                else -> OnLongPress
+            }
+            awaitVariant(initialDown, filter)
+        }
+
+    @OptIn(ExperimentalFoundationApi::class)
+    val OnSlop: suspend AwaitPointerEventScope.(PointerEvent, PointerFilter) -> DragStartResult? = { initialDown, _ ->
+        var overSlop = Offset.Zero
+        var drag: PointerInputChange?
+        do {
+            drag = awaitPointerSlopOrCancellation(
```

Comment:
It seems we also need to check `filter` here to cancel the drag. Without checking the filter we have a bug:

1. Press Right mouse on Left draggable
2. Press Left, release Left
3. Drag with the pressed right button - drag will happen, but shouldn't

It is difficult to inject this functionality into existing `awaitPointerSlopOrCancellation`, but we can run a separate coroutine, which checks cancelation. This will also help to move `filter` functionality out of this function into `draggable` itself. After that we can remove `PointerFilter` from this interface, and the users won't be able to write custom functions with `filter` violation.

Running a separate coroutine from `AwaitPointerEventScope` isn't possible, so we should run it outside of it and communicate via channel. Check [my branch](https://github.com/JetBrains/androidx/commit/9e32f467b66ec2cf8976332de1d6dcb2298478a2), where I tried to implement this.  It somewhat ugly and incomplete (we don't call onDragCancel). So our alternative is to switch to the old Config style (your example in Slack), or remove this parameter, and implement it somewhen in the future.

Reply:
done, took your implements  of communication via channel.
removed awaitForStart parameter (use slop for now for all pointer types)